[WIP] Mute organization badge and server count

kanishk98 opened the pull request:
---
<!--
Remove the fields that are not appropriate
Please include:
-->

**What's this PR do?**
Adds context menu and settings option to mute/unmute badge count of connected organizations. 

**Any background context you want to provide?**
https://github.com/zulip/zulip-electron/issues/623 should provide the required context. Please also have a look at the chat thread referenced.

**Screenshots?**
![image](https://user-images.githubusercontent.com/24617297/51103833-2e7a0900-180a-11e9-98d5-5e6a1a6dac41.png)
![image](https://user-images.githubusercontent.com/24617297/51103850-405bac00-180a-11e9-9129-9bc55624cfc6.png)

**You have tested this PR on:**
  - [x] Windows
  - [ ] Linux/Ubuntu
  - [ ] macOS

kanishk98 committed:
Add mute button to context menu and org settings

kanishk98 committed:
Save muted/unmuted status per org

Adds mutedOrganizations object to settings.json file inside the app's
data folder to keep track of muted organizations.

kanishk98 committed:
Mute badges and tray icons

Depending on per-org setting, the app will now hide the badge count of
certain servers.

Fix unmute/mute confirmation prompt

abhigyank commented on app/renderer/js/components/webview.js:
```diff
@@ -117,6 +117,11 @@ class WebView extends BaseComponent {
 	getBadgeCount(title) {
+		const mutedOrganizations = ConfigUtil.getConfigItem('mutedOrganizations');
+		const { url } = this.props;
+		if (!!mutedOrganizations[url]) {
```
Jus out of curiosity, any reason to use `!!mutedOrganizations[url]` instead of simply `mutedOrganizations[url]` ?

kanishk98 replied:
Not really, just the way I got used to. I'll change this along with the Travis fixes and squash the commits.

abhigyank reviewed (commented):
The feature works as expected and required. Good work on that.
A couple of suggestions with implementation though - 
* Instead of storing whether org is muted in configutil, I think i'd be better to store that as a property of the org in domain-util.json.
* I feel Muting/Unmuting a realm causing a refresh could be inconvenience to the user. I feel we should be able to do maybe to this without refreshing.
* Not sure about the design of the additions in the organisations page. Maybe @akashnimare  would weigh in on that.

kanishk98 commented:
> Instead of storing whether org is muted in configutil, I think i'd be better to store that as a property of the org in domain-util.json.

I attempted that initially by adding a `notify` flag to the `serverConf` object, but this approach didn't work out for `zulipdev.org` servers. The flag would get added and then removed from the file. This was before you helped me with the console logs; I'll let you know how this goes.

> I feel Muting/Unmuting a realm causing a refresh could be inconvenience to the user. I feel we should be able to do maybe to this without refreshing.

I agree. I'll get to work. :)

kanishk98 commented:
@abhigyank the latest commit addresses the refresh problem, but behaviour is undesirable in one case - if the user is on the org preferences page and then mutes an org using the context menu, it does not trigger an update of the DOM element in the pane. 
My initial intuition is that if we can keep track of where the user is in the app, then we can figure out a way to cause the update in the preferences page if the user mutes the organization using the context menu.

abhigyank commented:
@kanishk98 the unread count now doesn't change for me in any case. Now since we are not reloading the app, you would need explicitly remove the badge count of the div `server-tab-badge` of that icon. 

> if the user is on the org preferences page and then mutes an org using the context menu, it does not trigger an update of the DOM element in the pane.

Not sure how we can tackle this but I think we can make use of IPCRenderers 'forward-message' feature  here.

kanishk98 commented:
Ah, sorry. Should've checked thoroughly.

> Not sure how we can tackle this but I think we can make use of IPCRenderers 'forward-message' feature here.

Thanks! I'll check that out.

showell commented:
Some relevant discussion is here:

https://chat.zulip.org/#narrow/stream/101-design/topic/mute.20organization/near/683125

kanishk98 committed:
Removes need to refresh after muting orgs

Individually handles refreshing UI without refreshing the whole app
after user mutes/unmutes server for both context menu and org
preferences pane.
Also improves code readibility by removing !! operations.

Mute individual orgs without refreshing entire app

Fix Travis errors

Fix Travis errors

akashnimare commented:
@rishig anything else we should be doing here functionality wise?

rishig commented:
What does the setting currently do?
Also, I think the conclusion was that this should be implemented in the webapp?

rishig commented on app/renderer/js/main.js:
```diff
@@ -530,6 +533,29 @@ class ServerManagerView {
 					}
+				},
+				{
+					label: (mutedOrganizations[url] ? 'Unmute' : 'Mute') + ' organization',
```
Note that stuff like this is not going to work with translations

kanishk98 replied:
I'm sorry, I didn't think about that. I'll need to have a look at how translations are currently handled. Will get to it.

kanishk98 commented:
> What does the setting currently do?

The setting hides the number of unread messages in the sidebar and the tray icon for all muted organizations. 

> Also, I think the conclusion was that this should be implemented in the webapp?

If you see [this thread](https://chat.zulip.org/#narrow/stream/137-feedback/topic/Mute.20organizations), you'll see that the last few messages indicate that the feature is required in the desktop app too. One of the suggestions from the community (see https://github.com/zulip/zulip-electron/issues/623#issuecomment-454795980) is to extend this functionality to other environments, but this is certainly needed on the desktop. :)

rishig commented:
ah, I guess we should have been more explicit. Given that it needs to be reflected on all clients, the setting should be stored on the UserProfile object on the server, and controlled by something in the webapp's settings page. That single setting will control how we set unread counts on desktop/mobile/terminal/web.

akashnimare commented:
@kanishk98 any update on this? https://github.com/zulip/zulip/pull/12677 should be helpful I guess. @rishig would it make sense to add an option like the following to mute an org in the desktop app -

![image](https://user-images.githubusercontent.com/2263909/62974309-8d9e0b80-be35-11e9-80e6-2a51008e4fd0.png)

rishig commented:
yeah, for sure; maybe called "None".
Though I imagine the current setting solved most of the problem for most people.

kanishk98 commented:
I'm confused about the discussion here - given that people can now control their notifications across all clients from the web app and we read notifications from the favicon updates, I thought the conclusion here was to move such settings to the web app and continue as is with the desktop app.

kanishk98 commented:
If we're planning to also include a similar setting on the desktop app, how does that add more to what's on the web app right now?

akashnimare commented:
> yeah, for sure; maybe called "None".

@kanishk98 if we add this in the webapp then we can re-use it in the desktop app to add muting org feature.

rishig commented:
I also agree with Kanishk's intuition; I think we likely don't need a separate desktop app feature for this. 
The benefit of having settings in the webapp is that those settings will translate to mobile as well.
